wakatime_blender/heartbeat_queue.py

```diff
@@ -0,0 +1,178 @@
+        # grab the name of the directory
+        name = os.path.basename(os.path.dirname(filename))
+    else:
+        # cut away the (.blend) extension
+        name = os.path.splitext(filename)[0]
+        # remove (the full) path from the filename
+        name = os.path.basename(name)
+    # remove trailing characters (as configured in "Preferences")
+    name = name.rstrip(truncate_trail)
+    # tune project-name with pre- and postfix
+    name = f"{project_prefix}{name}{project_postfix}"
+    log(INFO, "project-name in Wakatime: {}", name)
+    return name
+
+
+@dataclass
+class HeartBeat:
+    entity: str
+    project: str
+    timestamp: float
+    is_write: bool = False
+
+
+class HeartbeatQueue(threading.Thread):
+    POLL_INTERVAL = 1
+
+    def __init__(self, version: str) -> None:
+        super().__init__()
+        self.daemon = True
+        self._version = version
+        self._queue = Queue()
+        self._last_hb: Optional[HeartBeat] = None
+        self._lock = threading.Lock()
+        self._running = True
+
+    def _enough_time_passed(self, now, is_write):
+        props = WakatimeProjectProperties.instance()
+        return self._last_hb is None or (
+            now - self._last_hb.timestamp
+            > (
+                2
+                if is_write
+                else min(1, (2 if not props else props.heartbeat_frequency) * 60)
+            )
+        )
+
+    def enqueue(self, filename: str, is_write=False):
+        timestamp = time.time()
+        last_file = self._last_hb.entity if self._last_hb is not None else ""
+        if (
+            not filename
+            or filename == last_file
+            and not self._enough_time_passed(timestamp, is_write)
+        ):
+            return
+        props = WakatimeProjectProperties.instance()
+        project_name = guess_project_name(
+            filename,
+            props.truncate_trail if props else "",
+            props.use_project_folder if props else False,
+            props.project_prefix if props else "",
+            props.project_postfix if props else "",
+        )
+        self._last_hb = HeartBeat(filename, project_name, timestamp, is_write)
+        self._queue.put_nowait(self._last_hb)
+
+    def shutdown(self):
+        self._queue.put_nowait(None)
+        with self._lock:
+            self._running = False
+
+    def _send_to_wakatime(
+        self, heartbeat: HeartBeat, extra_heartbeats: Optional[List[HeartBeat]] = None
+    ):
+        ua = f"blender/{bpy.app.version_string.split()[0]} blender-wakatime/{self._version}"
+        cmd = [
+            sys.executable,
+            settings.API_CLIENT,
+            "--entity",
+            heartbeat.entity,
+            "--time",
+            f"{heartbeat.timestamp:f}",
+            "--plugin",
+            ua,
+        ]
+        props = WakatimeProjectProperties.instance()
+        if props and props.always_overwrite_name:
+            cmd.extend(["--project", heartbeat.project])
```

Comment:
Usually we use `--alternate-project` to let wakatime-cli try detecting the project first. That keeps all the plugins consistent when they're working in the same project.

Reply:
I've described this behaviour in the REDME here:

https://github.com/allista/WakatimeBlender/blob/v2/README.md#configuration

The thing is, usually there's no project per se with blender, just a file. No git either. So this option is added specifically for cases when wakatime cannot detect the name of the de-facto absent project.

Reply:
In that case when no project is detected, `--alternate-project` works the same way as `--project` but alternate is safer.

Reply:
That's why the setting is called "always overwrite" and there's the `else:` clause below which does use `--alternate-project`. This gives a user more control over their statistics.